### PR TITLE
1482 Strict Parsing for Decimal

### LINKED ISSUE #1482: Add a strict parsing mode to Decimal type.
## Background
When a string is parsed to Decimal type, in case the string contains more decimal places then the Decimal's scale parameter the extra decimal places are trimmed.

## Feature
Implement a strict parsing option - triggered by metadata entry, defaulting to `false` - that will generate an error in `ErrCol` in case the decimal places overflow the scale of the define Decimal type.

## Example
Under strict rule the parsing would be as follows:
Decimal(10,2)
"123.45" -> 123.45, no error in `ErrCol`
"1234,567-> 1234.56, error in `ErrCol`

### PR DESCRIPTION
**Pull request by AdrianOlosutean**
Closes #1482

### CODE REVIEW

File: utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala [hunk 1]
```diff
@@ -349,14 +349,21 @@ object TypedStructField {
...
       }
 
     override def validate(): Seq[ValidationIssue] = {
-      NumericFieldValidator.validate(this)
+      DecimalFieldValidator.validate(this)
+    }
+
+    override protected def convertString(string: String): Try[BigDecimal] = {
+      if (strictParsing && scale < string.split("\\.")(1).length)
+          Failure(new IllegalArgumentException(s"Strict parsing does not allow $string"))
```
Comment (Zejnilovic) [inline utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala:365]: This message will need an update. It is not strict parsing that doesn't allow random values. It is the `s"Mismatch between the scale of the data and the schema defined scale resulting in loss of precission"` or something like this

Comment (AdrianOlosutean) [reply utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala:365]: True, the message should be more explanatory. How about `$string exceeds the defined scale limit in the schema`

Comment (Zejnilovic) [reply utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala:365]: Even better

File: utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala [hunk 2]
```diff
@@ -349,16 +349,25 @@ object TypedStructField {
...
+
+    override def needsUdfParsing: Boolean = strictParsing || super.needsUdfParsing
 
     override def validate(): Seq[ValidationIssue] = {
-      NumericFieldValidator.validate(this)
+      DecimalFieldValidator.validate(this)
     }
 
+    override protected def convertString(string: String): Try[BigDecimal] =
+      parser.flatMap(parser => parser.parse(string)).flatMap(_ => super.convertString(string))
```
Comment (benedeki) [inline utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala:369]: Why is this needed?

**Review by benedeki — COMMENTED** [review]
Also, please can you prepare a gh-pages update on the new metadata, please.

File: utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala (hunk 2 above)
Comment (AdrianOlosutean) [reply utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala:369]: True, it's redundant

File: utils/src/main/scala/za/co/absa/enceladus/utils/types/parsers/DecimalParser.scala [hunk 3]
```diff
@@ -16,11 +16,15 @@
...
+
 import za.co.absa.enceladus.utils.numeric.NumericPattern
 
+import scala.util.{Failure, Success, Try}
+
 class DecimalParser(override val pattern: NumericPattern,
                     override val min: Option[BigDecimal],
-                    override val max: Option[BigDecimal])
+                    override val max: Option[BigDecimal],
+                    val maxScale: Option[Int])
```
[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply utils/src/main/scala/za/co/absa/enceladus/utils/types/parsers/DecimalParser.scala:27]: You mean `None`?

[the remark this answers is not in the document]
Comment (benedeki) [reply utils/src/main/scala/za/co/absa/enceladus/utils/types/parsers/DecimalParser.scala:27]: Yup, sorry.

**Review by HuvarVer — APPROVED** [review]
I think useablility is good enough. Don't forget to update the documentation.
Done tests of new feature and smoke tests of UI, done e2e tests by Hermese 0.2.2

File: utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala [hunk 4]
```diff
@@ -349,16 +349,25 @@ object TypedStructField {
...
       DecimalTypeStructField.maxPossible(dataType)
     ){
+    val strictParsing: Boolean = getMetadataStringAsBoolean(MetadataKeys.StrictParsing).getOrElse(false)
 
-    override val parser: Try[DecimalParser] =
+    override val parser: Try[DecimalParser] = {
+      val maxScale = if(strictParsing) Some(scale) else None
       pattern.map { patternOpt =>
-        DecimalParser(patternOpt.getOrElse(NumericPattern(defaults.getDecimalSymbols)), Option(typeMin), Option(typeMax))
+        val pattern: NumericPattern = patternOpt.getOrElse(NumericPattern(defaults.getDecimalSymbols))
```
[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala:357]: Yes, and otherwise the line would be too long

File: utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala [hunk 5]
```diff
@@ -366,7 +366,7 @@ object TypedStructField {
     }
 
     override protected def convertString(string: String): Try[BigDecimal] =
-      parser.flatMap(parser => parser.parse(string)).flatMap(_ => super.convertString(string))
+      parser.flatMap(parser => parser.parse(string))
```
Comment (benedeki) [inline utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala:369]: This equals to inherited code. No need to override. 😉

**Review by benedeki — APPROVED** [review]

**Review by HuvarVer — APPROVED** [review]
